--- vte-0.14.1/configure.in.orig	2006-12-10 11:58:20.655981000 +0100
+++ vte-0.14.1/configure.in	2006-12-10 12:10:32.397680585 +0100
@@ -31,7 +31,7 @@
 	if test -d "$ac_x_includes" ; then
 		X_CFLAGS="-I$ac_x_includes $X_CFLAGS"
 	fi
-	X_LIBS="$X_PRE_LIBS $X_LIBS -lX11 $X_EXTRA_LIBS"
+	X_LIBS="$X_LIBS -lX11 $X_EXTRA_LIBS"
 	if test -d "$ac_x_libraries" ; then
 		X_LIBS="-L$ac_x_libraries $X_LIBS"
 	fi
@@ -266,9 +266,9 @@
 AC_CHECK_FUNC(floor,,AC_CHECK_LIB(m,floor,LIBS=["$LIBS -lm"]))
 
 # Look for ncurses or curses or termcap.
-AC_CHECK_HEADER(ncurses.h,[AC_CHECK_HEADER(term.h,[AC_CHECK_LIB(ncurses,tgetent,[LIBS="-lncurses $LIBS";AC_DEFINE(HAVE_NCURSES,1,[Define if you have ncurses.h and libncurses])])])])
-if test x$ac_cv_lib_ncurses_tgetent != xyes ; then
-	AC_CHECK_HEADER(ncurses/curses.h,[AC_CHECK_HEADER(ncurses/term.h,[AC_CHECK_LIB(ncurses,tgetent,[LIBS="-lncurses $LIBS";AC_DEFINE(HAVE_NCURSES_CURSES,1,[Define if you have ncurses/curses.h and libncurses])])])])
-	if test x$ac_cv_lib_ncurses_tgetent != xyes ; then
+AC_CHECK_HEADER(ncurses.h,[AC_CHECK_HEADER(term.h,[AC_CHECK_LIB(tinfo,tgetent,[LIBS="-ltinfo $LIBS";AC_DEFINE(HAVE_NCURSES,1,[Define if you have ncurses.h and libncurses])],[AC_CHECK_LIB(ncurses,tgetent,[LIBS="-lncurses $LIBS";AC_DEFINE(HAVE_NCURSES,1,[Define if you have ncurses.h and libncurses])])])])])
+if test x$ac_cv_lib_ncurses_tgetent != xyes -a x$ac_cv_lib_tinfo_tgetent != xyes ; then
+	AC_CHECK_HEADER(ncurses/curses.h,[AC_CHECK_HEADER(ncurses/term.h,[AC_CHECK_LIB(tinfo,tgetent,[LIBS="-ltinfo $LIBS";AC_DEFINE(HAVE_NCURSES_CURSES,1,[Define if you have ncurses/curses.h and libncurses])],[AC_CHECK_LIB(ncurses,tgetent,[LIBS="-lncurses $LIBS";AC_DEFINE(HAVE_NCURSES_CURSES,1,[Define if you have ncurses/curses.h and libncurses])])])])])
+	if test x$ac_cv_lib_ncurses_tgetent != xyes -a x$ac_cv_lib_tinfo_tgetent != xyes ; then
 		AC_CHECK_HEADER(curses.h,[AC_CHECK_HEADER(term.h,[AC_CHECK_LIB(curses,tgetent,[LIBS="-lcurses $LIBS";AC_DEFINE(HAVE_CURSES,1,[Define if you have curses.h and libcurses])])])])
 		if test x$ac_cv_lib_curses_tgetent != xyes ; then
