--- vte-0.22.1/src/vteseq.c~	2009-09-24 21:25:51.000000000 +0200
+++ vte-0.22.1/src/vteseq.c	2009-09-28 00:34:06.418540003 +0200
@@ -1060,7 +1060,7 @@
 			rowdata = _vte_ring_index_writable (screen->row_data, i);
 			g_assert(rowdata != NULL);
 		} else {
-			rowdata = _vte_terminal_ring_append (screen->row_data, FALSE);
+			rowdata = _vte_terminal_ring_append (terminal, FALSE);
 		}
 		/* Pad out the row. */
 		_vte_row_data_fill (rowdata, &screen->fill_defaults, terminal->column_count);
