diff -ur vte-0.11.11.orig/configure.in vte-0.11.11/configure.in
--- vte-0.11.11.orig/configure.in	2004-05-02 08:43:01.000000000 +0200
+++ vte-0.11.11/configure.in	2004-08-16 16:20:48.546759640 +0200
@@ -378,13 +378,13 @@
 # If we can find libzvt on the system, set up to build a variant of "reflect"
 # which uses it.
 REFLECT_ZVT=false
-if pkg-config --exists libzvt-2.0 ; then
-	REFLECT_ZVT=true
-	ZVT_DEFS=-DHAVE_ZVT
-	PKG_CHECK_MODULES(ZVT,[gtk+-2.0 libzvt-2.0])
-fi
-AC_SUBST(ZVT_DEFS)
-AM_CONDITIONAL(REFLECT_USE_ZVT, $REFLECT_ZVT)
+#if pkg-config --exists libzvt-2.0 ; then
+#	REFLECT_ZVT=true
+#	ZVT_DEFS=-DHAVE_ZVT
+#	PKG_CHECK_MODULES(ZVT,[gtk+-2.0 libzvt-2.0])
+#fi
+#AC_SUBST(ZVT_DEFS)
+#AM_CONDITIONAL(REFLECT_USE_ZVT, $REFLECT_ZVT)
 
 mylibdir=`eval echo $libdir`
 mylibexecdir=`eval echo $libexecdir`
