diff -aurN vte-0.13.1.orig/src/keymap.c vte-0.13.1/src/keymap.c
--- vte-0.13.1.orig/src/keymap.c	2006-04-19 00:36:42.000000000 +0200
+++ vte-0.13.1/src/keymap.c	2006-06-09 22:49:21.981672736 +0200
@@ -203,16 +203,14 @@
  * system to system, or mine's just broken.  But anyway. */
 static const struct _vte_keymap_entry _vte_keymap_GDK_Home[] = {
 	{cursor_all, keypad_all, fkey_all, 0, X_NULL, 0, "kh"},
-	{cursor_all, keypad_all, fkey_vt220, 0, _VTE_CAP_CSI "1~", -1, X_NULL},
+	{cursor_all, keypad_all, fkey_vt220, 0, _VTE_CAP_SS3 "H", -1, X_NULL},
 	{cursor_all, keypad_all, fkey_hp, 0, _VTE_CAP_ESC "h", -1, X_NULL},
-	{cursor_all, keypad_all, fkey_nothpvt, 0, X_NULL, 0, "kh"},
 	{cursor_all, keypad_all, fkey_all, 0, X_NULL, 0, X_NULL},
 };
 
 static const struct _vte_keymap_entry _vte_keymap_GDK_End[] = {
 	{cursor_all, keypad_all, fkey_all, 0, X_NULL, 0, "@7"},
-	{cursor_all, keypad_all, fkey_vt220, 0, _VTE_CAP_CSI "4~", -1, X_NULL},
-	{cursor_all, keypad_all, fkey_notvt220, 0, X_NULL, 0, "@7"},
+	{cursor_all, keypad_all, fkey_all, 0, _VTE_CAP_SS3 "F", -1, X_NULL},
 	{cursor_all, keypad_all, fkey_all, 0, X_NULL, 0, X_NULL},
 };
 
@@ -999,6 +997,13 @@
 	*special = NULL;
 	*normal_length = 0;
 
+	/* Map Shift-F1 to F11 ... Shift-F10 to F20 */
+	if (keyval >= GDK_F1 && keyval <= GDK_F10 &&
+	    (modifiers & GDK_SHIFT_MASK)) {
+		keyval += 10;
+		modifiers &= ~GDK_SHIFT_MASK;
+	}
+
 	/* Search for the list for this key. */
 	entries = NULL;
 	for (i = 0; i < G_N_ELEMENTS(_vte_keymap); i++) {
@@ -1251,7 +1256,7 @@
 _vte_keymap_key_gets_modifiers(guint keyval)
 {
 	gboolean fkey = FALSE;
-	/* Determine if this key gets modifiers. */
+	/* Determine if the key can take modifiers. */
 	switch (keyval) {
 	case GDK_Up:
 	case GDK_Down:
@@ -1259,6 +1264,8 @@
 	case GDK_Right:
 	case GDK_Insert:
 	case GDK_Delete:
+	case GDK_Home:
+	case GDK_End:
 	case GDK_Page_Up:
 	case GDK_Page_Down:
 	case GDK_KP_Up:
