--- vte-0.74.0/src/vte.cc.orig	2023-09-21 06:28:59.064532927 +0200
+++ vte-0.74.0/src/vte.cc	2023-09-21 06:28:53.223547680 +0200
@@ -98,7 +98,7 @@ static inline double round(double x) {
 }
 #endif /* !HAVE_ROUND */
 
-#define WORD_CHAR_EXCEPTIONS_DEFAULT "-#%&+,./=?@\\_~\302\267"sv
+#define WORD_CHAR_EXCEPTIONS_DEFAULT "-#%&+,./=?@:;\\_~\302\267"sv
 
 #define I_(string) (g_intern_static_string(string))
 
